pkg/scd/store/cockroach/store.go

```diff
@@ -13,33 +15,85 @@ var (
 	DefaultClock = clockwork.NewRealClock()
 )
 
-// Store is an implementation of dss.Store using
-// Cockroach DB as its backend store.
+// Store is an implementation of scd.Store using
+// a CockroachDB transaction.
 type Store struct {
-	*cockroach.DB
+	tx     *sql.Tx
+	ctx    context.Context
 	logger *zap.Logger
 	clock  clockwork.Clock
 }
 
-// NewStore returns a Store instance connected to a cockroach instance via db.
-func NewStore(db *cockroach.DB, logger *zap.Logger) *Store {
+// Transaction is an implementation of scd.Transaction using
+// a CockroachDB transaction.
+type Transaction struct {
+	tx     *sql.Tx
+	logger *zap.Logger
+	ctx    context.Context
+	clock  clockwork.Clock
+}
+
+// Implement store.Transaction interface
+func (t *Transaction) Store() (scdstore.Store, error) {
 	return &Store{
-		DB:     db,
+		tx:     t.tx,
+		logger: t.logger,
+		ctx:    t.ctx,
+		clock:  t.clock,
+	}, nil
+}
+
+// Implement store.Transaction interface
+func (t *Transaction) Commit() error {
+	return t.tx.Commit()
+}
+
+// Implement store.Transaction interface
+func (t *Transaction) Rollback() error {
+	return t.tx.Rollback()
+}
+
+// Transactor is an implementation of scd.Transactor using
+// a CockroachDB database.
+type Transactor struct {
```

Comment:
Have you taken a look at what I did in the RID refactor? This is very similar, and we should probably align the approaches.

In particular the other approach solves some of the following:

1. Protects against incorrectly nested transactions.
2. Uses the crdb-go library for retries which retries more types of retry-able errors.
3. Builds in timeouts for the retries
4. One struct instead of the 2 here (Store and Transactor).

Any reason in particular you're diverging from that setup here?

Reply:
This PR was started before 6997658070f2b98dbc100a665a9ce015c8759188 and the [documentation](https://github.com/interuss/dss/blob/master/pkg/rid/README.md#todo) seems out of sync if the TODOs being resolved already.

Re: 1, Could you elaborate on how/in what circumstances this implementation would have a problem with nested transactions?

Re: 3, Is there a structural reason why retries couldn't be added relatively easy to this architecture?

Re 4, It seems like a Repository from RID is an overloaded struct -- it represents both a repo outside a transaction and a repo inside a transaction.  When it is a repo outside a transaction, it seems dangerous to use most of its methods as the developer could easily not realize they are performing operations on the repo outside a transaction.  When a repo is inside a transaction, then the InTxnRetrier method doesn't really make sense to have, nor does Close.  The approach here (which is otherwise pretty similar, as you mentioned) fixes this issue by making the roles of the structs clear: the Transactor only has the equivalent of InTxnRetrier and Store only has the equivalent of the substantive operations.

Reply:
ah, took a closer look i see that nested txn's aren't possible.

Re 4: I disagree, in fact there's many queries here we wouldn't want to use transactions with, so it gives the developer the flexibility to do the right thing. Forcing transactions on the user is not what we want.

Splitting them out into 2 would be ok, but only for the reason of losing the Close and Begin functions within a transaction.